bug fixes

spicy-sauce opened the pull request:
autopep8 formatter - max_line_length option is ignored #4
recv() doesn't guarantee receiving the full message #6

spicy-sauce committed:
autopep8 formatter - max_line_length option is ignored #4

spicy-sauce committed:
recv() doesn't guarantee receiving the full message #6

spicy-sauce commented:
#6 is buggy, do not merge just yet.

spicy-sauce commented:
Closing this for now, I'll open a new PR once fixing #6.

spicy-sauce committed:
recv() doesn't guarantee receiving the full message #6

spicy-sauce committed:
Updated version

spicy-sauce commented:
Fixed, please go over and approve.
Also, please update pip accordingly.

guyskk commented on pybeautifier.py:
```diff
@@ -152,6 +154,26 @@ def serve():
+            chunk = client.recv(4096)  # some 2^n number
+            if not chunk:
+                break
+            data += chunk.decode('utf-8')
```
Should decode after received all data bytes, otherwise decode non-complete bytes may failed.

spicy-sauce replied:
Good catch, thanks.

guyskk commented on pybeautifier.py:
```diff
@@ -152,6 +154,26 @@ def serve():
+                break
+            data += chunk.decode('utf-8')
+
+        except socket.error as ex:
```
seems not needed, exceptions are handled in L146.

spicy-sauce replied:
Indeed, thanks.

guyskk commented on pybeautifier.py:
```diff
@@ -152,6 +154,26 @@ def serve():
         server.close()
 
 
+def read_message(client) -> str:
```
How about use below code? not need use select if read from only one socket.
```python
data = b''
while True:
    chunk = client.recv(4096)
    if not chunk:
        break
    data += chunk
return data.decode('utf-8')
```

spicy-sauce replied:
This may get into an infinite loop... 

How about this one?
`def read_message(client) -> str:
    BUFF_SIZE = 4096
    data = b''
    while True:
        chunk = client.recv(BUFF_SIZE)
        if not chunk:
            break

        data += chunk
        if len(chunk) < BUFF_SIZE:
            break

    return data.decode('utf-8')`

guyskk replied:
It will not get infinite loop: if chunk is not empty then it will receive at least 1 byte, until received all data. right?

spicy-sauce replied:
Checked it once more, this might get into an infinite loop.

Please, see this one:
```
def read_message(client) -> str:
    BUFF_SIZE = 4096
    data = b''
    while True:
        chunk = client.recv(BUFF_SIZE)
        if not chunk:
            break
                        	
        data += chunk
        if len(chunk) < BUFF_SIZE:
            break

    return data.decode('utf-8')

```

Makes sense?

spicy-sauce replied:
debugging it some more...

spicy-sauce replied:
OK, it won't get into an infinite loop, but in both our code snippets, in case the last chunk's length equals to the buffer size, the next invocation of client.recv will hangs as there's no more bytes to read, see this:
https://stackoverflow.com/questions/25495593/when-using-socket-recv-the-program-hangs-up-when-there-are-no-more-bytes-to-r

This doesn't happen when using select.

guyskk replied:
> in case the last chunk's length equals to the buffer size, the next invocation of client.recv will hangs as there's no more bytes to read.

Yes. The client process should shutdown write after send all request data, so that server will receive an empty chunk.

And `len(chunk) < BUFF_SIZE` not indicate no more data, for example: client send 1 byte, wait 10 seconds, send another byte, then the server may receive two 1 byte chunk.

guyskk replied:
Use this API: https://nodejs.org/docs/latest/api/net.html#net_socket_end_data_encoding_callback

And my [old client code](https://github.com/Glavin001/atom-beautify/blob/41d5051b399f7fe11efd149504a232df16179d60/src/beautifiers/pybeautifier.coffee#L25) is also wrong.

guyskk replied:
The book **Unix Network Programming** is very good, I learned most socket programing from it. But this project was write before I read the book, so it is not robust and has some bugs.

The 1 connection per request mode is not very suitable for online service, it cost more connection time, and TCP socket programing is a bit complex. I suggest built HTTP service instead, if you need high performance and robust online service.

spicy-sauce replied:
Yes, this is the plan, providing a HTTP service to achieve this goal.
I've decided to use pybeautifier at this point since its capabilities are sufficient for my current needs, but I should change it (probably in a new repo, but we can improve pybeautifier if it's OK with you).

So, for now, we should fix the current 2 issues (#4 and #6 - if the change suggested by you in the client does the trick, I'll replace the use of select with the alternative).

guyskk replied:
For #4 the fix is good and can merge. 
For #6, fix the server will broken the auto-beautify plugin, but I don't have time to change it.

I suggest you fork this repo, and release your distro to PYPI.

spicy-sauce replied:
As you suggested, it seems that having client.end() right after invoking the client.write() solves the issue and the pybeautifier related code can be:
```
def read_message(client) -> str:
    data = b''
    while True:
        chunk = client.recv(4096)
        if not chunk:
            break
        data += chunk

    return data.decode('utf-8')
```

That means that in atom-beautify you should just add the client.end() after line 25, right?
I can go with forking the repo, or just add this to atom-beautify so we'll have it working.
Let me know what you think so we can finalize this either way.

Thanks!

guyskk replied:
`That means that in atom-beautify you should just add the client.end() after line 25, right?`
Yes, you are right. You can create PR to atom-beautify, then I will merge the fix for #6.

If you don't want to maintain atom-beautify plugin then you should fork this repo.

spicy-sauce replied:
Fixed #6 as discussed (please review).
Opened a new PR to atom-beautify:
https://github.com/Glavin001/atom-beautify/pull/2578

guyskk replied:
Thank you!

spicy-sauce committed:
recv() doesn't guarantee receiving the full message #6

- Decode only after receiving all data bytes, otherwise decode non-complete bytes may failed.
- Removed redundant exception handling
- read_message - replaced select impl.
